### PR TITLE
[CL]: Fix tick rounding bug and implement direct conversion from tick <> sqrt price

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #5516

> **Note to reviewer:** the only real state machine change is in `tick.go` and heavily mirrors @ValarDragon's PR here: #5522 
> The rest of the changes are related to function renames/test refactors.

## What is the purpose of the change

This PR expands on #5522 and updates all price to tick conversions to use SqrtPriceToTick instead.

## Testing and Verifying

The new function is tested in `math/tick_test.go`

The original attack vector is also converted into an edge case test in `position_test.go`

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

File: x/concentrated-liquidity/keeper_test.go [hunk 1]
```diff
@@ -49,10 +49,10 @@ var (
...
 	DefaultTickSpacing                                   = uint64(100)
 	PoolCreationFee                                      = poolmanagertypes.DefaultParams().PoolCreationFee
-	DefaultExponentConsecutivePositionLowerTick, _       = math.PriceToTickRoundDown(sdk.NewDec(5500), DefaultTickSpacing)
-	DefaultExponentConsecutivePositionUpperTick, _       = math.PriceToTickRoundDown(sdk.NewDec(6250), DefaultTickSpacing)
-	DefaultExponentOverlappingPositionLowerTick, _       = math.PriceToTickRoundDown(sdk.NewDec(4000), DefaultTickSpacing)
-	DefaultExponentOverlappingPositionUpperTick, _       = math.PriceToTickRoundDown(sdk.NewDec(4999), DefaultTickSpacing)
+	DefaultExponentConsecutivePositionLowerTick, _       = math.PriceToTickRoundDownSpacing(sdk.NewDec(5500), DefaultTickSpacing)
+	DefaultExponentConsecutivePositionUpperTick, _       = math.PriceToTickRoundDownSpacing(sdk.NewDec(6250), DefaultTickSpacing)
+	DefaultExponentOverlappingPositionLowerTick, _       = math.PriceToTickRoundDownSpacing(sdk.NewDec(4000), DefaultTickSpacing)
+	DefaultExponentOverlappingPositionUpperTick, _       = math.PriceToTickRoundDownSpacing(sdk.NewDec(4999), DefaultTickSpacing)
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/keeper_test.go:55]: Driveby rename as the original name could be misinterpreted to refer to rounding behavior

File: x/concentrated-liquidity/swaps.go [hunk 2]
```diff
@@ -365,8 +365,7 @@ func (k Keeper) computeOutAmtGivenIn(
 		} else if !sqrtPriceStart.Equal(sqrtPrice) {
 			// Otherwise if the sqrtPrice calculated from computeSwapStep does not equal the sqrtPrice we started with at the
 			// beginning of this iteration, we set the swapState tick to the corresponding tick of the sqrtPrice calculated from computeSwapStep
-			price := sqrtPrice.Mul(sqrtPrice)
-			swapState.tick, err = math.PriceToTickRoundDown(price, p.GetTickSpacing())
+			swapState.tick, err = math.SqrtPriceToTickRoundDownSpacing(sqrtPrice, p.GetTickSpacing())
```
Comment (p0mvn) [inline x/concentrated-liquidity/swaps.go:368]: Unsure whether in/out of the scope of this PR but we should not be doing tick spacing rounding in swaps, only in LP logic to prevent initialization of ticks we do not desire.

It should be possible to be within tick spacing for the "current tick" upon swap completion. However, it should not be possible to initialize the tick at an index that is not a multiple of tick spacing.

Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/swaps.go:368]: Hmm thanks for flagging this – not sure I fully follow the concern but sounds important. Let's discuss separately

File: x/concentrated-liquidity/position_test.go [hunk 3]
```diff
@@ -2506,27 +2506,53 @@ func (s *KeeperTestSuite) TestCreateFullRangePositionLocked() {
...
+	desiredTokenOut := sdk.NewCoin(USDC, sdk.NewInt(10000))
+	s.FundAcc(swapAddr, sdk.NewCoins(sdk.NewCoin(ETH, sdk.NewInt(1000000000000000000))))
+	_, _, _, _, _, err := s.clk.SwapInAmtGivenOut(s.Ctx, swapAddr, pool, desiredTokenOut, ETH, sdk.ZeroDec(), sdk.ZeroDec())
+	s.Require().NoError(err)
+
+	// Both positions should be able to withdraw successfully
+	_, _, err = s.clk.WithdrawPosition(s.Ctx, firstPositionAddr, firstPosId, firstPosLiq)
+	s.Require().NoError(err)
+	_, _, err = s.clk.WithdrawPosition(s.Ctx, secondPositionAddr, secondPosId, secondPosLiq)
+	s.Require().NoError(err)
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/position_test.go:2535]: Should we assert any invariance here on the amount withdrawn?

File: x/concentrated-liquidity/range_test.go [hunk 4]
```diff
@@ -149,7 +149,8 @@ func (s *KeeperTestSuite) setupRangesAndAssertInvariants(pool types.Concentrated
 
 	// Ensure the pool balance is exactly equal to the assets added + amount swapped in - amount swapped out
 	poolAssets := s.App.BankKeeper.GetAllBalances(s.Ctx, pool.GetAddress())
-	s.Require().Equal(totalAssets, poolAssets)
+	poolSpreadRewards := s.App.BankKeeper.GetAllBalances(s.Ctx, pool.GetSpreadRewardsAddress())
+	s.Require().Equal(totalAssets, poolAssets.Add(poolSpreadRewards...))
```
Comment (czarcas7ic) [inline x/concentrated-liquidity/range_test.go:153]: Do we need to add the pool's incentive address here as well?

File: x/concentrated-liquidity/swaps.go (hunk 2 above)
Comment (czarcas7ic) [reply x/concentrated-liquidity/swaps.go:368]: Not sure if I am following here, do you mean a pools current tick inter swap can be between tick spacings or the result can be between tick spacings? I would think at the very least the resulting tick once the swap is complete should be rounded.

**Review by czarcas7ic — COMMENTED** [review]
Well done on this, will approve after some discourse in some of these comments

File: x/concentrated-liquidity/range_test.go (hunk 4 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/range_test.go:153]: I don't believe so – the reason we have to add spread rewards is because the tracking we do wrt swaps is inclusive of fees (we add `tokenIn` and subtract `tokenOut` from `totalAssets` after each swap). Incentives aren't part of this equation

File: x/concentrated-liquidity/position_test.go (hunk 3 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/position_test.go:2535]: We already cover this in the invariant fuzz tests so this behavior should already be thoroughly covered. I mostly just added this case as a formality to show that the original attack vector no longer triggered a panic

File: x/concentrated-liquidity/math/tick.go [hunk 5]
```diff
@@ -204,9 +227,85 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
...
+	truncatedTick := tick.TruncateInt64()
+
+	// We have a candidate bucket index `t`. We discern here if:
+	// * sqrtPrice in [TickToSqrtPrice(t - 1), TickToSqrtPrice(t))
+	// * sqrtPrice in [TickToSqrtPrice(t), TickToSqrtPrice(t + 1))
+	// * sqrtPrice in [TickToSqrtPrice(t+1), TickToSqrtPrice(t + 2))
+	// * sqrtPrice not in either.
+	// We handle boundary checks, by saying that if our candidate is the min tick,
+	// set the candidate to min tick + 1.
+	// If our candidate is the max tick, set the candidate to max tick - 2.
```
[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:271]: It's minus 2 because to check tick t + 1, we need to go to t + 2. Thus, to not go over max tick during these checks, we need to shift down by 2

File: x/concentrated-liquidity/invariant_test.go [hunk 6]
```diff
@@ -144,10 +144,10 @@ func (s *KeeperTestSuite) assertWithdrawAllInvariant() {
 		totalWithdrawn = totalWithdrawn.Add(withdrawn...)
 	}
 
-	// We allow for an additive tolerance of 1 per position in the pool, since this is the maximum that can be truncated
-	// when collecting spread rewards.
+	// We allow for an additive tolerance of 2 per position in the pool, since we expect up to 1 to be lost to rounding on
+	// join and withdrawal each.
 	errTolerance := osmomath.ErrTolerance{
-		AdditiveTolerance: sdk.NewDec(int64(len(allPositions))),
+		AdditiveTolerance: sdk.NewDec(2 * int64(len(allPositions))),
```
[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/invariant_test.go:150]: This is a good point. Investigating now.

[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/invariant_test.go]: Discussed offline and converged on relaxing the invariant to simply check for rounding direction instead of specifying an exact additive amount, as we slightly overround in parts of our logic and it is not worth blocking this PR on changing this

File: x/concentrated-liquidity/math/tick.go [hunk 7]
```diff
@@ -204,9 +227,85 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
...
+	outOfBounds := false
+	if truncatedTick <= types.MinTick {
+		truncatedTick = types.MinTick + 1
+		outOfBounds = true
+	} else if truncatedTick >= types.MaxTick-1 {
+		truncatedTick = types.MaxTick - 2
+		outOfBounds = true
```
[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:278]: Err might be missing your point – if truncated tick = MaxTick, it enters this branch since MaxTick >= MaxTick - 1, no?

[the remark this answers is not in the document]
Comment (czarcas7ic) [reply x/concentrated-liquidity/math/tick.go:278]: You are right, was just not clear on first pass, thanks

**Review by czarcas7ic — APPROVED** [review]
Spoke offline IRT that unanswered questions. Approving but looking for a second approval before merge. Nice work again :)

File: x/concentrated-liquidity/swaps.go (hunk 2 above)
Comment (p0mvn) [reply x/concentrated-liquidity/swaps.go:368]: Yes, that's what I mean. Current tick can be between two initialized ticks inter-swap. That is, inside an active bucket. With tick rounding, the translation between current sqrt price and current tick would be broken.

I addressed this concern in this PR: https://github.com/osmosis-labs/osmosis/pull/5526 and left more detailed explanation.

Let's continue discussing this in #5526

File: x/concentrated-liquidity/position_test.go (hunk 3 above)
Comment (p0mvn) [inline x/concentrated-liquidity/position_test.go:2535]: Commenting for visibility, I expect this to solve this issue in #5526 : 
https://github.com/osmosis-labs/osmosis/pull/5526/files#r1233117388

File: x/concentrated-liquidity/math/tick.go [hunk 8]
```diff
@@ -204,9 +227,85 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
...
+func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
+	// TODO: Make truncate, since this defines buckets as
+	// [TickToPrice(b - .5), TickToPrice(b+.5))
+	return calculatePriceToTickDec(price).RoundInt64()
+}
+
+// CalculatePriceToTick takes in a square root and returns the corresponding tick index.
+// This function does not take into consideration tick spacing.
+// NOTE: This is really returning a "Bucket index". Bucket index `b` corresponds to
+// all sqrt prices in range [TickToSqrtPrice(b), TickToSqrtPrice(b+1)).
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/tick.go:256]: Per definitions here:
https://github.com/osmosis-labs/osmosis/blob/bd809a973132d55a63e7c34c326ec1408b38eec2/x/concentrated-liquidity/README.md?plain=1#L1589-L1596

I would not call this a bucket but, instead, a tick. The reason why I think these semantics are important is that a tick can be within a bucket.

File: x/concentrated-liquidity/math/tick.go [hunk 9]
```diff
@@ -113,38 +114,64 @@ func PriceToTick(price sdk.Dec) (int64, error) {
...
+}
+
 // PriceToTickRoundDown takes a price and returns the corresponding tick index.
 // If tickSpacing is provided, the tick index will be rounded down to the nearest multiple of tickSpacing.
+// NOTE: This should not be called in the state machine.
 // CONTRACT: tickSpacing must be smaller or equal to the max of 1 << 63 - 1.
 // This is not a concern because we have authorized tick spacings that are smaller than this max,
 // and we don't expect to ever require it to be this large.
-func PriceToTickRoundDown(price sdk.Dec, tickSpacing uint64) (int64, error) {
+func PriceToTickRoundDownSpacing(price sdk.Dec, tickSpacing uint64) (int64, error) {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/concentrated-liquidity/math/tick.go:150]: +1. If we are set on the new approach, I would like to suggest removing the old broken logic.

I understand that it will require refactoring tests and can be done in a separate PR. However, I think it would be a time well spent as a few months later the context will be lost, and someone will start wondering why we have two kinds of price to tick conversions.

For this PR, I think we should at least move it outside of the core logic like Adam suggested

File: x/concentrated-liquidity/math/tick.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/concentrated-liquidity/math/tick.go:271]: Could you please add this context to the comments?

**Review by p0mvn — COMMENTED** [review]
Great work! I would like to do another pass tomorrow and run tests more. Leaving nits for now

File: x/concentrated-liquidity/math/tick.go (hunk 8 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:256]: Hmm yeah it seems the definition has shifted slightly (this was a comment dev originally added). I just removed the note to avoid further confusion

File: x/concentrated-liquidity/math/tick.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go]: Of course. Added here: 2a4f1c4cd6e3c4b14da70dceeb0fbf92eaa45faa

File: x/concentrated-liquidity/math/tick.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go]: Discussed offline and decided to move into apptesting as a short term fix to get these functions out of the state machine: 08bd83f286f1ab2ecd2b493717b95569eb73acb1

File: x/concentrated-liquidity/math/tick.go [hunk 10]
```diff
@@ -204,9 +227,85 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
...
+// NOTE: This is really returning a "Bucket index". Bucket index `b` corresponds to
+// all prices in range [TickToSqrtPrice(b), TickToSqrtPrice(b+1)).
+// We make an erroneous assumption here, that bucket index `b` corresponds to
+// all prices in range [TickToPrice(b), TickToPrice(b+1)).
+// This currently makes this function unsuitable for the state machine.
+func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
+	// TODO: Make truncate, since this defines buckets as
+	// [TickToPrice(b - .5), TickToPrice(b+.5))
+	return calculatePriceToTickDec(price).RoundInt64()
+}
```
[the remark this answers is not in the document]
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go]: 08bd83f286f1ab2ecd2b493717b95569eb73acb1

File: app/apptesting/concentrated_liquidity.go [hunk 11]
```diff
@@ -141,3 +143,57 @@ func (s *KeeperTestHelper) SetupConcentratedLiquidityDenomsAndPoolCreation() {
 	defaultParams.AuthorizedQuoteDenoms = append(defaultParams.AuthorizedQuoteDenoms, ETH, USDC, BAR, BAZ, FOO, UOSMO, STAKE)
 	s.App.ConcentratedLiquidityKeeper.SetParams(s.Ctx, defaultParams)
 }
+
+// PriceToTickRoundDown takes a price and returns the corresponding tick index.
+// If tickSpacing is provided, the tick index will be rounded down to the nearest multiple of tickSpacing.
+// CONTRACT: tickSpacing must be smaller or equal to the max of 1 << 63 - 1.
+// This is not a concern because we have authorized tick spacings that are smaller than this max,
+// and we don't expect to ever require it to be this large.
+func (s *KeeperTestHelper) PriceToTickRoundDownSpacing(price sdk.Dec, tickSpacing uint64) (int64, error) {
```
Comment (AlpinYukseloglu) [inline app/apptesting/concentrated_liquidity.go:152]: Note to reviewer: these functions were moved to be high level test helpers to get them out of the state machine without breaking the many tests that rely on them to (correctly) generate expected values

File: x/concentrated-liquidity/range_test.go [hunk 12]
```diff
@@ -47,7 +47,7 @@ type RangeTestParams struct {
 var (
 	DefaultRangeTestParams = RangeTestParams{
 		// Base amounts
-		baseNumPositions:     1000,
+		baseNumPositions:     10,
```
Comment (AlpinYukseloglu) [inline x/concentrated-liquidity/range_test.go:50]: Note to reviewer: the changes in this file are mainly cleanup/test speed related changes to ensure tests run in a reasonable amount of time

File: x/concentrated-liquidity/math/tick.go [hunk 13]
```diff
@@ -204,9 +227,85 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
...
+	}
+
+	// We error if sqrtPriceT is above sqrtPriceTplus2 or below sqrtPriceTmin1.
+	// For cases where calculated tick does not fall on a limit (min/max tick), the upper end is exclusive.
+	// For cases where calculated tick falls on a limit, the upper end is inclusive, since the actual tick is
+	// already shifted and making it exclusive would make min/max tick impossible to reach by construction.
+	// We do this primary for code simplicity, as alternatives would require more branching and special cases.
+	if (!outOfBounds && sqrtPrice.GTE(sqrtPriceTplus2)) || sqrtPrice.LT(sqrtPriceTmin1) || sqrtPrice.GT(sqrtPriceTplus2) {
+		return 0, fmt.Errorf("sqrt price to tick could not find a satisfying tick index. Hit bounds: %v", outOfBounds)
+	}
```
[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/concentrated-liquidity/math/tick.go:242]: I am confused about the `!outOfBounds` conditional as tied to this

File: x/concentrated-liquidity/position_test.go [hunk 14]
```diff
@@ -2506,27 +2506,53 @@ func (s *KeeperTestSuite) TestCreateFullRangePositionLocked() {
...
+		"two adjacent ranges": {
+			tickRanges: [][]int64{
+				{-10000, 10000},
+				{10000, 20000},
+			},
+			rangeTestParams: DefaultRangeTestParams,
+		},
 
 		// Both of these lead to underclaiming of fees greater than additive
 		// error tolerance of 1 per token per position. Increasing ticks increases
```
Comment (ValarDragon) [inline x/concentrated-liquidity/position_test.go:2558]: does this case still error after this PR?

File: x/concentrated-liquidity/math/tick.go [hunk 15]
```diff
@@ -204,9 +185,73 @@ func CalculatePriceToTick(price sdk.Dec) (tickIndex int64) {
...
+	// max tick during these checks, we need to shift it down by 2.
+	// We check this at max tick - 1 instead of max tick, since we expect the output to
+	// have some error that can push us over the tick boundary.
+	outOfBounds := false
+	if truncatedTick <= types.MinTick {
+		truncatedTick = types.MinTick + 1
+		outOfBounds = true
+	} else if truncatedTick >= types.MaxTick-1 {
+		truncatedTick = types.MaxTick - 2
+		outOfBounds = true
```
Comment (ValarDragon) [inline x/concentrated-liquidity/math/tick.go:224]: I wonder if its simpler to just deal with these by checking the boundary cases in a separate function, on the sqrt price directly. (pre-squaring)

File: x/concentrated-liquidity/math/tick_test.go [hunk 16]
```diff
@@ -709,8 +709,114 @@ func (suite *ConcentratedMathTestSuite) TestCalculatePriceToTick() {
...
+			sqrtPrice:    sdk.MustNewDecFromStr("9999999499999987499.999374999960937496"),
+			tickSpacing:  1,
+			tickExpected: types.MaxTick - 2,
+		},
+		"sqrt price one ULP below max tick - 1 (tick spacing 100)": {
+			// Calculated using TickToSqrtPrice(types.MaxTick - 1) - 1 ULP
+			sqrtPrice:    sdk.MustNewDecFromStr("9999999499999987499.999374999960937496"),
+			tickSpacing:  defaultTickSpacing,
+			tickExpected: types.MaxTick - defaultTickSpacing,
+		},
```
Comment (p0mvn) [inline x/concentrated-liquidity/math/tick_test.go:801]: Should we also be having test cases around min tick bounds similar to how this is done for max tick?

File: x/concentrated-liquidity/position_test.go (hunk 14 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/position_test.go:2558]: No since we relaxed the invariant checks to allow for slight overrounding recently. Will uncomment

**Review by p0mvn — APPROVED** [review]
LGTM. Great work

File: x/concentrated-liquidity/math/tick.go (hunk 15 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick.go:224]: Possibly, although my previous attempts at this led to constraining the "real" min tick and max tick by 1-2 ticks (which might be fine but leads to more cases that need to be checked/handled)

If you see a way to abstract this logic without triggering this, would be great for us to do a follow up PR to clean this up!

File: x/concentrated-liquidity/math/tick_test.go (hunk 16 above)
Comment (AlpinYukseloglu) [reply x/concentrated-liquidity/math/tick_test.go:801]: Will add these cases in the min tick change PR as the expected values are thrown off due to canonical tick rounding
